include/openssl/params.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef HEADER_PARAMS_H
+# define HEADER_PARAMS_H
+
+# include <openssl/core.h>
+# include <openssl/bn.h>
+
+# ifdef  __cplusplus
+extern "C" {
+# endif
+
+# define OSSL_PARAM_DEFN(key, type, addr, sz, rsz)    \
+    { (key), (type), (addr), (sz), (rsz) }
+
+# define OSSL_PARAM_int32(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(int32_t), NULL)
+# define OSSL_PARAM_uint32(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(uint32_t), NULL)
+# define OSSL_PARAM_int64(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(int64_t), NULL)
+# define OSSL_PARAM_uint64(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(uint64_t), NULL)
+# define OSSL_PARAM_intmax(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_INTEGER, (addr), sizeof(intmax_t), NULL)
+# define OSSL_PARAM_uintmax(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), \
+                    sizeof(uintmax_t), NULL)
+# define OSSL_PARAM_size_t(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), sizeof(size_t), \
+               NULL)
+# define OSSL_PARAM_double(key, addr) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_REAL, (addr), sizeof(double), NULL)
```

Comment:
These macros are useful for setting up a OSSL_PARAM array to pass to a *setter* function (`set_params`).  There should be a matching set for setting up a OSSL_PARAM array to pass to a *responder* function (`get_params`).  The difference?  That last parameter, the return size pointer.

Reply:
I'm not sure I understand fully.  These macros are designed to make allocating an array of params easier.  I.e.

```
static const OSSL_PARAM my_definitions[] = {
    OSSL_PARAM_double("entropy", &entropy_param),
    OSSL_PARAM_uint64("bytes", &num_bytes),
    { 0 }
};
```

They are not meant to be directly used for either getting or setting.  The functions do those.

I do need to add a terminating macro though.

Reply:
So let's say that some provider has a parameter "foo" documented like this:

> "foo" is an unsigned integer, 32 bits in size on I32 systems (such as x86), 64 bits on I64 systems (such as x86_64)

Possible result:

``` C
    uint64_t foo = 0;
    size_t foo_l = 0;
    OSSL_PARAM params[] = {
        { "foo", OSSL_PARAM_UNSIGNED_INTEGER, &foo, sizeof(foo), &foo_l },
        { 0 }
    };

    whatever_get_params(obj, params);
#ifdef B_ENDIAN
    if (foo_l == sizeof(uint32_t)) {
        foo = foo >> 32;  /* quick cheap and dirty narrowing */
#endif 
```

Of course, there are a gazillion ways to do this, but the question remains if the provider and the application have the same idea of if "foo" should have a 32-bit or a 64-bit value.  In such cases, *the return size is of interest*

Either way, some application writers may be interested in the return size either way, just as a safety measure.

Reply:
We will need macros that allow a return size to be specified as well.

Do we need support for variable sized integers?  Rich has included _int_ and _long_ which are variable.  I didn't but could (in addition to the fixed sized values).  I don't see having additional types as being a problem.

Reply:
I see no problem adding support for `int` and `long`...  It's not like they really are variable size *within an architecture*...

Reply:
They are added.

Reply:
The return size macros are there now too.